docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,192 @@
+# Dead letter queue
+
+## Introduction
+
+In Conduit the concept of a Dead letter queue (DLQ) describes the place where
+messages are sent if they are negatively acknowledged (nacked). This can happen
+for a number of reasons:
+
+- A node in the pipeline experiences an error while processing a message (e.g.
+  processor).
+- The destination connector fails to process a message and sends back a nack.
+- The pipeline is stopped forcefully causing in-flight messages to be nacked.
+
+### Goals
+
+The purpose of a DLQ is to make a pipeline resilient to errors that might
+happen while processing messages.
+
+The design for the DLQ should support the following:
+
+- Allow user to reroute nacked messages to any Conduit connector (builtin or
+  standalone).
+- A message sent to the DLQ should reflect the contents it had just before the
+  error happened. Example: if a message went through 2 processors and failed
+  while being processed by a 3rd processor, the DLQ message should reflect any
+  transforms done by the first 2 processors.
+- Ensure message order is preserved when messages are sent to the DLQ.
+  In other words - the order in which messages were produced by the source
+  connector should be preserved, even if messages are nacked in a different
+  order (this is already assured using
+  a [semaphore](https://github.com/ConduitIO/conduit/blob/b6325584d51ec64bc9086faddcdaa2788a5dfa8f/pkg/pipeline/stream/source_acker.go#L35-L37)).
+- If a message failed to be sent to the DLQ the nack action should fail and
+  stop the pipeline.
+- Allow the user to configure the window and number of messages that need to be
+  nacked to consider the pipeline broken and trigger a stop. Example: if 25 of
+  the last 30 messages are nacked stop the pipeline.
+- Messages pushed to the DLQ should contain metadata describing the error and
+  the source connector.
+
+### Questions
```

Comment:
Should we allow multiple DLQs? I imagine there would be use cases such as nacked messages from one source go into the same DLQ, or that there are DLQs for records of different importance levels.

Reply:
I honestly don't know how many users would need this, that's why I sided on the side of simplicity, so we don't work on something that potentially won't be used by anyone. That said, the workaround to achieve this behavior would be to sort records after they land in the DLQ. For example, you could push DLQ messages into a certain Kafka topic and have another pipeline that reads that topic, categorizes the messages and moves them to different DLQ topics.

Reply:
Even though I can imagine use cases, I can't say either how many would use, so I definitely agree: let's keep it simple.:)